Adding interface for refactoring.

navch opened the pull request:
This PR is just to add interface for Code refactoring. This have same architecture as CodeHintManager. To provide some refactoring features (i.e. Rename, Extract etc) we can register a provider(as extension) directly. This interface will take care of adding commands and menus for all available handlers. This also responsible for calling appropriate handler based on priority.

What's next - 
----------------
- [ ] This interface should also take care of some common UI pieces for all languages

Upcoming PR's which are going to use this interface
- [ ] Rename reference in scope for JavaScript
- [ ] Extract some expression to a variable in scope for JavaScript

Note :- This PR Uses submenu API @boopeshmahendran raise a [PR](https://github.com/adobe/brackets/pull/13824) to provide addSubmenu API.

Please Review @swmitra @petetnt @nethip @boopeshmahendran

Commit:
Adding interface for refactoring. So for different languages we can add extension directly

petetnt commented:
Cool stuff @navch! I'll try to review this soon-ish.

nethip commented on src/editor/CodeRefactoringManager.js:
```diff
@@ -0,0 +1,231 @@
+ * collection of refactoring providers. If user opens context menus 
+ * then the providers registered for the current language are queried
+ * for their ability to provide refactoring handlers in order of descending priority by
+ * way their hasHanlders methods.
```
nit: correct the spelling mistake.

nethip replied:
I think there are quite a few spelling mistakes in the description. Could you correct all of these?

nethip commented on src/editor/CodeRefactoringManager.js:
```diff
@@ -0,0 +1,231 @@
+     * The method by which a CodeRefactoringProvider registers its willingness to
+     * help in code refactoring for editors in a given language.
+     *
+     * @param {!providerInfo} provider
```
Can use mention the structure of `providerInfo`?

nethip commented on src/editor/CodeRefactoringManager.js:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        // Find a suitable provider, if any
+        var language = editor.getLanguageForSelection(),
```
How expensive is this operation `getLanguageForSelection`?

nethip commented on src/editor/CodeRefactoringManager.js:
```diff
@@ -0,0 +1,231 @@
+        editorCmenu = Menus.getContextMenu(Menus.ContextMenuIds.EDITOR_MENU);
+
+        // Check for handlers on right click
+        editorCmenu.on("beforeContextMenuOpen", function (e) {
```
Check `editorCmenu` for null?

nethip commented on src/editor/CodeRefactoringManager.js:
```diff
@@ -0,0 +1,231 @@
+     * Find a suitable provider for a language id and
+     * Add a context menu if there is any provider available
+     */
+    function getHandlersForLanguageId() {
```
You might want to change the name of this function as it more than a `getter`. Also all private function need to have `_` prepended.

nethip commented:
@navch Thanks for this PR! Here is my feedback

* My view is that the framework seems to be too generic. You might want to club common operations as common commands and let the extension handle the language specific part of Rename. You can think of it as three common commands exposed, and the later ask the extensions to add in for more. Otherwise from the UX point of view, it is a distraction for me to know the order and understand the menus when we switch between languages like HTML/JS/CSS.

* I am not sure having the refactor part only in sub-menus is a good option. If some extension overwrites the context menu then there is no way refactor can ever be used. How about having those options in the top menu as well? I know that we don't yet support sub menus but it is a good to give it a thought in this perspective.

* Have an option to associate a shortcut with the rename and refactoring menus is a good idea.

* Apart from the back-end part, there is no help w.r.t UI paradigm for extension developers. It might be a good thing, in terms of freedom but also bad thing in terms of different providers implementing rename and refactoring differently for different languages. We might want to still stick with this framework but come up with a Wiki page as to document the most common practices.


@petetnt Did you get a chance to review this PR?

navch commented:
Closing this PR for now.  Later on, if we plan to ship refactoring feature for other languages then we can open this again after addressing these review comments.
